### PR TITLE
Feature: Add Chrome History 

### PR DESCRIPTION
**Pull request by ganyariya**
### Add Chrome History Feature

Hello, I would like to add the feature "Browser History Search".
We sometimes want to search by `History`, therefore I'd like to add this feature.

### What I did

- Add `Interfaces` about `Browser History` , referencing `Browser Bookmark` implementation.
- Add `GoogleChromeHistoryRepository`

The histories of Google Chrome are [managed by sqlite](https://gist.github.com/dropmeaword/9372cbeb29e8390521c2).
And, Chrome locks the sqlite file of history while using Google chrome.
Therefore, I copy sqlite file as `UeliHistory`.

### What I still do not

- Add Unit Test Cases
- Settings Interface of history
- `Sqlite Lock Problem` better solution (copy is not great solution, I think... If use copy sqlite file, ueli must copy file periodically, because history goes alone by using chrome by user.)

I would try to fix and refine codes and resolve above problems.
So, I am very happy to give me review, and advise.
I am pleased to fix and add codes.
(So, please do not merge :)  )

Thank you!

### CODE REVIEW

File: src/main/plugins/browser-histories-plugin/google-chrome-history-repository.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public copyDatabaseFile(): void {
+        fs.copyFile(this.originalDatabaseFilePath, this.ueliDatabaseFilePath, (err) => {
+            if (err) {
+                throw err;
+            }
+        });
+    }
+
```
Comment (Ganariya) [inline src/main/plugins/browser-histories-plugin/google-chrome-history-repository.ts:42]: Copy part, so I'd like to refine this

**Comment by oliverschwendener** [comment]
Thanks alot for your contribution! I'll have a quick look at the code.

File: src/main/plugin-type.ts [hunk 2]
```diff
@@ -20,6 +20,7 @@ export enum PluginType {
     ColorConverter = "color-converter",
     Dictionary = "dictionary",
     BrowserBookmarks = "browser-bookmarks",
+    BrowserHistories = "browser-histories",
```
Comment (oliverschwendener) [inline src/main/plugin-type.ts:23]: Why don't we call this feature `BrowserHistory`? Makes more sense in my opinion

File: src/main/plugins/browser-histories-plugin/browser-history.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export interface BrowserHistory {
```
Comment (oliverschwendener) [inline src/main/plugins/browser-histories-plugin/browser-history.ts:1]: I'd call this interface `BrowserHistoryEntry` or something similar. As far as I understand the "history" is a collection of entries, and this interface represents the single entry, right?

File: src/main/plugin-type.ts (hunk 2 above)
Comment (Ganariya) [reply src/main/plugin-type.ts:23]: Thank you for your review!
I think so too, 'history' refers to `records`, and 'historyentry' refers to one record.

I'll fix these problems and refine some codes and features!

File: src/main/plugins/browser-histories-plugin/browser-histories-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
+import { SearchPlugin } from "../../search-plugin";
+import { PluginType } from "../../plugin-type";
+import { SearchResultItem } from "../../../common/search-result-item";
+import { UserConfigOptions } from "../../../common/config/user-config-options";
+import { TranslationSet } from "../../../common/translation/translation-set";
+import { BrowserHistory } from "./browser-history";
+import { BrowserHistoryOptions } from "../../../common/config/browser-histories-options";
+import { BrowserHistoryRepository } from "./browser-history-repository";
+
+export class BrowserHistoriesPlugin implements SearchPlugin {
```
[the remark this answers is not in the document]
Comment (Ganariya) [reply src/main/plugins/browser-histories-plugin/browser-histories-plugin.ts:10]: Thank you! I'll fix this name method!

File: src/main/plugins/browser-history-plugin/browser-history-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        this.browserHistory = [];
+    }
+
+    public async getAll(): Promise<SearchResultItem[]> {
+        return await this.browserHistory.map((history) => this.buildSearchResultItem(history));
+    }
+
+    public async refreshIndex(): Promise<void> {
+        const history = await this.getBrowserHistory();
+        this.browserHistory = history;
```
Comment (oliverschwendener) [inline src/main/plugins/browser-history-plugin/browser-history-plugin.ts:38]: How about simplifiying this?

```typescript
this.browserHistory = await this.getBrowserHistory();
```

File: src/main/plugins/browser-history-plugin/browser-history-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    ) {
+        this.config = config;
+        this.translations = translations;
+        this.browserHistoryRepositories = browserHistoryRepositories;
+        this.urlExecutor = urlExecutor;
+        this.browserHistory = [];
+    }
+
+    public async getAll(): Promise<SearchResultItem[]> {
+        return await this.browserHistory.map((history) => this.buildSearchResultItem(history));
```
Comment (oliverschwendener) [inline src/main/plugins/browser-history-plugin/browser-history-plugin.ts:33]: I think this await is redundant.

File: src/main/plugins/browser-history-plugin/browser-history-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public async refreshIndex(): Promise<void> {
+        const history = await this.getBrowserHistory();
+        this.browserHistory = history;
+    }
+
+    public async clearCache(): Promise<void> {
+        return new Promise((resolve) => resolve());
```
Comment (oliverschwendener) [inline src/main/plugins/browser-history-plugin/browser-history-plugin.ts:42]: Can we simplify this?

```typescript
public async clearCache(): Promise<void> {
    return;
}
```

**Comment by oliverschwendener** [comment]
When I checked out your branch the app won't start anymore? Can you look into this? Maybe you have to merge latest `dev` back to your branch?

**Comment by oliverschwendener** [comment]
I will close this PR due to inactivity. Feel free to open this up again if you decide to continue the work ;)

**Pull request closed without merging**